doc/design/eventdriven-callbacks.md

```diff
@@ -0,0 +1,215 @@
+      of the API and bugs caused by misuse.
+ - (Reactor pattern)[https://en.wikipedia.org/wiki/Reactor_pattern]
+ - (Condition Variable)[https://en.wikipedia.org/wiki/Monitor_(synchronization)#Condition_variables_2]
+ 
+## Design
+
+### General
+The usage should be similar to the _WaitSet_ with a key difference - it should 
+be **event driven** and not a mixture of event and state driven, depending on
+which event is attached, like in the _WaitSet_.
+
+It should have the following behavior:
+
+ - Whenever an event occurs the corresponding callback should be called **once**
+    immediately.
+ - If an event occurs multiple times before the callback was called, the callback
+    should be called **once**.
+ - If an event occurs while the callback is being executed the callback should be 
+    called again **once**.
+
+The following use cases and behaviors should be implemented.
+
+ - The API and ReactAl should be robust this means that it should be impossible 
+    to use the API in the wrong way. Since ReactAl is working concurrently a 
+    wrong usage could lead to race conditions, extremely hard to debug bug reports 
+    (HeisenBugs) and can be frustrating to the user.
+    We list here features marked with [robust] which are only supported to 
+    increase this kind of the robustness.
+
+ - concurrently: attaching a callback at any time from anywhere. This means in particular
+    - Attaching a callback from within a callback.
+    - [robust] Updating a callback from within the same callback 
+      ```cpp
+      void sampleReceived2(iox::popo::UntypedSubscriber & subscriber) {}
+
+      void sampleReceived(iox::popo::UntypedSubscriber & subscriber) {
+        myCallbackReactal.attachEvent(subscriber, iox::popo::SubscriberEvent::HAS_SAMPLE_RECEIVED, sampleReceived2);
+      }
+
+      myCallbackReactal.attachEvent(subscriber, iox::popo::SubscriberEvent::HAS_SAMPLE_RECEIVED, sampleReceived);
+      ```
+ - concurrently: detach a callback at any time from anywhere. This means in particular
+    - Detaching a callback from within a callback.
+    - [robust] When the detach call returns we guarantee that the callback is never called
+      again. E.g. blocks till the callback is removed, if the callback is concurrently 
+      running it will block until the callback is finished and removed.
+      Exception: If a callback detaches itself it blocks until the callback is removed 
+                 and not until the callback is finished.
+    - Calling detach means that after that call the callback is no longer attached
+      even when it was not attached in the first place. Therefore it will always succeed.
+
+ - When the ReactAl goes out of scope it should detach itself from every class 
+     to which it was attached via a provided callback (like in the WaitSet).
+
+ - When the class which is attached to the ReactAl goes out of scope it should 
+    detach itself from the ReactAl via a provided callback (like in the WaitSet).
```

Comment:
Does it make sense to have a common base class for the WaitSet and ReactAl?

Reply:
Maybe?! But I would not go into so much detail in the first design draft since this will the implementation later show. But I created a task to update the design document when the implementation is finished so that when we get a base class it is documented here.